File: go/array/56_merge.go
package array

import "sort"

// merge 合并区间
func merge_2(intervals [][]int) [][]int {

	// 排序intervals
	sort.Sort(sortedArray(intervals))

	length := len(intervals)

	for i := 0; i < length-1; {
		cur, next := intervals[i], intervals[i+1]

		if cur[1] >= next[0] {
			if cur[1] < next[1] {
				intervals[i][1] = intervals[i+1][1]
			}

			intervals = append(intervals[:i+1], intervals[i+2:]...)
			length--
			continue
		}
		i++
	}

	return intervals

}

type sortedArray [][]int

func (s sortedArray) Len() int {
	return len(s)
}

func (s sortedArray) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

func (s sortedArray) Less(i, j int) bool {
	if s[i][0] == s[j][0] {
		return s[i][1] <= s[j][1]
	}
	return s[i][0] <= s[j][0]
}

File: rust/src/array/56_merge.rs
pub fn merge(intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
    let mut intervals = intervals;
    // 排序
    intervals.sort();
    let mut ans = vec![];

    let (mut l, mut r) = (intervals[0][0], intervals[0][1]);

    intervals.iter().skip(1).for_each(|x| {
        // 后面的元素大于前面的元素 不需要合并
        if x[0] > r {
            ans.push(vec![l, r]);
            l = x[0]
        }
        r = r.max(x[1]);
    });
    ans.push(vec![l, r]);
    ans
}
